src/Development/IDE/Core/Rules.hs

```diff
@@ -253,51 +252,51 @@ getLocatedImportsRule =
-                                in ((pathMap', ss'), (imp, Just pathId))
-                    -- Convert paths in imports to ids and update the path map
-                    let ((pathIdMap, ss'), modImports') = mapAccumL f (rawPathIdMap rawDepInfo, ss) modImports
-                    -- Files that we haven’t seen before are added to the queue.
-                    let newFiles =
-                            IntSet.fromList (coerce $ mapMaybe snd modImports')
-                            IntSet.\\ IntMap.keysSet (rawImports rawDepInfo)
-                    let rawDepInfo' = insertImport fId (Right $ ModuleImports modImports' pkgImports) rawDepInfo
-                    go (newFiles `IntSet.union` fs)
-                       (rawDepInfo' { rawPathIdMap = pathIdMap }, ss')
-
-
+    go :: [NormalizedFilePath] -> (RawDependencyInformation, IntMap ArtifactsLocation)
+                               -> Action ([FilePathId], RawDependencyInformation, IntMap ArtifactsLocation)
+    go [] (r, i) = pure ([], r, i)
+    go (f:fs) a@(rawDepInfo, ss)
+      | Just fid <- lookupPathToId (rawPathIdMap rawDepInfo) f = do
+          (fids, r, bms) <- go fs a
+          return (fid: fids, r, bms)
+      | otherwise = do
+          al <- modSummaryToArtifactsLocation f <$> use_ GetModSummary f
+          importsOrErr <- use GetLocatedImports f
+          let (fId, path_map) = getPathId al (rawPathIdMap rawDepInfo)
+          let ss' = if isBootLocation al
+                      then IntMap.insert (getFilePathId fId) al ss
+                      else ss
+          let rawDepInfo' = rawDepInfo { rawPathIdMap = path_map }
+          case importsOrErr of
+            Nothing -> do
+            -- File doesn’t parse
+              let rawDepInfo'' = insertImport fId (Left ModuleParseError) rawDepInfo'
+              (fids, r1, r2) <- go fs (rawDepInfo'', ss)
+              return (fId: fids, r1, r2)
+            Just (modImports, pkgImports) -> do
+              -- Get NFPs of the imports which have corresponding files
+              let (no_file, with_file) = splitImports modImports
+                  (mns, ls) = unzip with_file
+              (fids, d', ss'') <- go (map artifactFilePath ls) (rawDepInfo', ss')
+              let moduleImports' = map (,Nothing) no_file ++ zipEqual "raw_dep" mns (map Just fids)
+              let rawDepInfo'' = insertImport fId (Right $ ModuleImports moduleImports' pkgImports) d'
+              (fids, r1, r2) <- go fs (rawDepInfo'', ss'')
+              return (fId : fids, r1, r2)
+
+    splitImports :: [(Located ModuleName, Maybe ArtifactsLocation)]
+                 -> ([Located (ModuleName)], [(Located ModuleName, ArtifactsLocation)])
+    splitImports [] = ([], [])
+    splitImports ((m, k):is) = let (ns, ls) = splitImports is
+                               in case k of
+                                    Nothing -> (m:ns, ls)
+                                    Just a  -> (ns, (m, a):ls)
```

Comment:
```suggestion
    splitImports = foldr splitImportsLoop ([],[])
    
    splitImportsLoop (import, Nothing) (ns,_) = (import:ns, ls)
    splitImportsLoop (import, Just artifact) (_, ls) = (ns, (import,artifact) : ls)
```

Reply:
I'd go for `partitionEithers . map (\(m,k) -> maybe (Left m) (\k -> Right (m, k)) k)`

Reply:
I find @pepeiborra’s suggestion to be the easiest to understand.